mautic/mautic
# Upgrade dependencies (BC breaks) in preparation for PHP 8.0 support

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | features
| Bug fix?                               | no
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | yes
| Automated tests included?              | no
| Related user documentation PR URL      | N/A
| Related developer documentation PR URL | N/A
| Issue(s) addressed                     | Fixes #N/A




#### Description:

Updates/upgrades many of Mautic's dependencies in preparation of adding PHP 8.0 support. Although PHP 8.0 support can be added after Mautic 4 has been released, some packages have BC breaks which we already need to tackle now.

Most notably, this PR includes a refactor of `joomla/http` (`mautic.http.connector`) to Guzzle (`mautic.http.client`). More details have been added to `UPGRADE-4.0.md`.

**We had to temporarily disable Rector since it is incredibly slow and throwing dozens of errors on newer versions, which will need more investigation later**. Also lots of `Can't handle this situation`. Example run where you can see this: https://github.com/mautic/mautic/runs/2213233300



#### Things that need testing
- Pretty much all of Mautic, but since we plan to include this PR in 4.0 Alpha we'll be testing extensively anyway.

#### Some blockers for PHP 8.0 support moving forward

Though we've already updated many dependencies in preparation for PHP 8.0, some of them don't support 8.0 yet. The good news is that most of them work on 8.0 already but don't explicitly declare support in their `composer.json` files. Here's an overview of dependencies that have 8.0 support in progress, and which we can move to a separate issue later for follow-up:

- `friendsofsymfony/oauth2-php` is not PHP 8.0-ready. Turns out we're lucky and someone just created a PHP 8 PR: https://github.com/FriendsOfSymfony/oauth2-php/pull/125 - looks like no changes are needed on our side apart from just bumping the dependency
- `joomla/filter` already seems to support PHP 8.0 (their test suite runs against it), but it doesn't explicitly declare support in their `composer.json`. Created a PR for that: https://github.com/joomla-framework/filter/pull/46
- `noxlogic/ratelimit-bundle` has PHP 8.0 support in progress. Looks like no changes from our side are required apart from updating the dependency when it's ready: https://github.com/jaytaph/RateLimitBundle/pull/113
- `mautic/transifex` already works on PHP 8.0, just had to declare explicit support for it: https://github.com/mautic/Transifex-API/pull/3
- `friendsofsymfony/oauth-server-bundle` is working on a new major version, but it's not finished yet, so we can't update just yet. However, they have removed PHP templating support in favor of Twig. Since Mautic still heavily relies on PHP templates, we wouldn't be able to update this dependency once we work on PHP 8.0 support. Created a PR to re-add the templating engine: https://github.com/FriendsOfSymfony/FOSOAuthServerBundle/pull/676

#### Steps to test this PR:
1. Checkout this PR locally
2. Ensure that `composer install` works and that you're able to access Mautic's dashboard
3. A code review would be good, but is challenging given the amount of changes and `joomla/http` refactor. The refactoring specifically took place in https://github.com/mautic/mautic/pull/9833/commits/bb0b6c15f3407e9a72b5fbd9ca5dcbca7e38c4da
